core/include/vecmem/memory/details/unique_alloc_deleter.hpp

```diff
@@ -10,12 +10,116 @@
 
 #include <type_traits>
 
-#include "vecmem/memory/details/unique_alloc_deleter_impl.hpp"
+#include "vecmem/memory/memory_resource.hpp"
+#include "vecmem/vecmem_core_export.hpp"
 
 namespace vecmem::details {
+/**
+ * @brief A deleter class for trivial allocations.
+ *
+ * This class provides all the necessary functionality to allow unique pointers
+ * to deallocate allocations of trivial types, as well as arrays of them.
+ *
+ * @tparam T The type to deallocate.
+ *
+ * @note This deleter class supports non-array types, array types of unknown
+ * bound, but not array types of known bound.
+ *
+ * @note This deleter class only supports types that are both trivially
+ * constructible and trivially destructible.
+ */
 template <typename T>
-using unique_alloc_deleter = std::enable_if_t<
-    std::is_trivially_constructible_v<std::remove_extent_t<T>> &&
-        std::is_trivially_destructible_v<std::remove_extent_t<T>>,
-    unique_alloc_deleter_impl>;
-}
+struct unique_alloc_deleter {
+    /*
+     * Similarly, we cannot destroy objects, so we need to ensure that
+     * deallocation of objects is semantically equivalent to their destruction.
+     */
+    static_assert(std::is_trivially_destructible_v<std::remove_extent_t<T>>,
+                  "Allocation pointer type must be trivially destructible.");
```

Comment:
C++20 concepts would be the perfect fit here.

Reply:
I know... :sob: